Deadchat Plays SS13: Ghost Control of Mobs

lewcc opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## What Does This PR Do
Ported from https://github.com/tgstation/tgstation/pull/55314.

Allows admins to give deadchat control over just about anything (immovable rods, singularities, Ian, your local clown) in classic Twitch-Plays style.
Control can be set in anarchy or democracy mode, with anarchy letting every move count as soon as possible (though putting users on individual cooldowns), and democracy picking the most popular option after a cooldown.

To answer your (likely) questions:
- You can only control the thing when you're orbiting it as a ghost.
- Your control messages *will not go to dchat*, and other people won't see them.
- Likewise, other messages about controlled objects have made will only be relayed to orbiters.
- Admins can start it by **a new option to grant deadchat control in the VV dropdown**, or by proc-calling `deadchat_plays`, with arguments for the mode and cooldown.
- this control can be ended similarly, a VV dropdown option to end control will appear on anything that's being dchat controlled.

<!-- Include a small to medium description of what your PR changes. -->
<!-- Document all changes, as not doing this may delay reviews or even discourage maintainers from merging your PR! -->
<!-- If your PR fixes an issue, add "Fixes #1234" somewhere in the PR description. This will automatically close the bug upon PR submission. -->

## Why It's Good For The Game
It's fun! I think it's a fun way to appease deadchat for a little by giving them something productive to do instead of salting. Also, it could make some events a bit more interesting, if you ask me. I've added some support for humans, so punpun could be let loose on the station under deadchat's control, for example. This could also be expanded down the line into a midround or something.
<!-- Add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Images of changes
Controlling a human:
https://user-images.githubusercontent.com/89928798/223902032-c40d7f7d-e3e0-404a-bd02-c812caf93b69.mp4

Controlling a rod:
https://user-images.githubusercontent.com/89928798/223902104-8347f150-a58a-4642-b9e5-ffb4acfcf385.mp4


<!-- If you did not make a map or sprite edit, you may delete this section. You may include a gif or mp4 of your feature if you want. -->

## Testing
Lots of controlling mobs.
<!-- How did you test the PR, if at all? -->

## Changelog
:cl:
add: Admins can now set items up to be controlled by deadchat!
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->
<!-- If a PR has no impact on players (i.e. a code refactor that does not change functionality) then the entire Changelog heading and contents can be removed. -->

lewcc commented on code/modules/mob/living/simple_animal/friendly/dog.dm:
```diff
@@ -105,6 +105,48 @@
+		"speak" = CALLBACK(src, .proc/handle_automated_speech, TRUE),
+		"wear_hat" = CALLBACK(src, .proc/find_new_hat),
+		"drop_hat" = CALLBACK(src, .proc/drop_hat),
+		"spin" = CALLBACK(src, /mob.proc/emote, "spin")), cooldown, CALLBACK(src, PROC_REF(end_dchat_plays)))
```
Fixed in https://github.com/ParadiseSS13/Paradise/pull/20600/commits/1a677564d60e0557b133341ae58e6dbb9a497f27

AffectedArc07 commented on code/datums/components/deadchat_control.dm:
```diff
@@ -0,0 +1,243 @@
+///Will execute a single command after the cooldown based on player votes.
+#define DEMOCRACY_MODE (1<<0)
```
Undef these please, or if you need them in other files, prefix them with ghostcontrol or something.

lewcc replied:
Addressed, moved em to out to a defines file as well

lewcc commented:
https://user-images.githubusercontent.com/89928798/224117656-22e6728e-fd80-42b9-a23a-2c0451981fb1.mp4


Added a vv menu item to make admemery a bit easier (note I've cleaned it up since taking this video).

AffectedArc07 commented on code/__DEFINES/misc_defines.dm:
```diff
@@ -538,3 +538,12 @@
+///Allows each player to do a single command every cooldown.
+#define DEADCHAT_ANARCHY_MODE (1<<1)
+///Mutes the democracy mode messages send to orbiters at the end of each cycle. Useful for when the cooldown is so low it'd get spammy.
+#define MUTE_DEMOCRACY_MESSAGES (1<<2)
```
```suggestion
#define MUTE_DEATCHAT_DEMOCRACY_MESSAGES (1<<2)
```

AffectedArc07 commented on code/datums/datumvars.dm:
```diff
@@ -1236,6 +1236,55 @@
 		CMT.ui_interact(usr)
 
+	if(href_list["grantdeadchatcontrol"])
+		if(!check_rights(R_ADMIN))
+			return
```
```suggestion
		if(!check_rights(R_EVENT))
			return
```

lewcc commented:
> This might be overcomplicating it, but it would be cool to see a list of votecounts for the current vote cycle somewhere on screen maybe?

Maybe? I could see that being cool if done right. I think think that'd be something worth looking into once the base functionality is in, though.

Probablyjay19 commented:
Can't wait to create year long rivalries with fellow spessmen due to 6 people not deciding on wether or not we should eat a bar of soap as a Dchat controlled drask.

Sirryan2002 commented on code/datums/components/deadchat_control.dm:
```diff
@@ -0,0 +1,240 @@
+		inputs[result].Invoke()
+		if(!(deadchat_mode & MUTE_DEADCHAT_DEMOCRACY_MESSAGES))
+			var/message = "<span class='deadsay italics bold'>[parent] has done action [result]!<br>New vote started. It will end in [input_cooldown * 0.1] second\s.</span>"
+			for(var/mob/dead/observer/M in orbiters)
+				to_chat(M, message)
```
will this send message to people with dsay disabled? If so, it needs to be able to be filtered due to players possibly not wanting their textbox spammed with a low-delay dchat plays ss13 game.

lewcc replied:
It should only send to people who are directly orbiting the object. That being said, I can definitely account for that on all these to_chats.

lewcc replied:
Addressed, added some checks to account for deadchat mutes.

Sirryan2002 reviewed (changes requested):
needs to be

lewcc commented on code/modules/power/singularity/singularity.dm:
```diff
@@ -421,7 +421,7 @@
 		M.visible_message("<span class='danger'>[M] stares blankly at [src]!</span>", \
 						"<span class='userdanger'>You look directly into [src] and feel weak.</span>")
-	return 
+	return
```
This wasn't one of my (intended) changes, was just an automatic whitespace fixes. There are a few other weird quirks with the returns in this file, might be something to look at later.

Sirryan2002 replied:
you can just resolve this then

lewcc replied:
It actually got removed when merging master anyway, haha

Sirryan2002 reviewed (approved):
honestly probably doesn't even need a TM since this is active only by admin intervention (Which means we can just not use it if it breaks). As long as this has been well-testing locally I'm fine with approving it
